PyOptionalOutputToolExample/PyOptionalOutputToolExampleEngine.py

```diff
@@ -0,0 +1,168 @@
+        self.message_type = Et.fromstring(str_xml).find('messageType').text
+        self.message_string = Et.fromstring(str_xml).find('messageString').text
+
+        # Getting the output anchor from Config.xml by the output connection name
+        self.output_anchor = self.output_anchor_mgr.get_output_anchor('Output')
+
+        return True
+
+    def pi_add_incoming_connection(self, str_type, str_name):
+        """
+        Called when the Alteryx engine is attempting to add an incoming data connection.
+        :param str_type: The type of each input connection.
+        :param str_name: A unique name for each input connection.
+        :return: Self, and a reference to an object.
+        """
+        return self
+
+    def pi_add_outgoing_connection(self, str_name):
+        """
+        Called when the Alteryx engine is attempting to add an outgoing data connection.
+        :param str_name: A unique name for each output connection.
+        :return: Boolean, where True signifies that the connection is accepted.
+        """
+        return True
+
+    def pi_push_all_records(self, n_record_limit):
+        """
+        Called when the Alteryx engine when it's expecting the plugin to provide all of its data.
+        Only pertinent to tools which have no upstream connections, like the Input tool.
+        :param n_record_limit: Set it to <0 for no limit, 0 for no records, and >0 to specify the number of records.
+        :return: False, prevent sending all data downstream.
+        """
+        self.alteryx_engine.output_message(self.n_tool_id, self.message.error, 'Missing Incoming Connection')
+        return False
+
+    def pi_close(self, b_has_errors):
+        """
+        Called after all data has finished flowing through all the fields.
+        :param b_has_errors: Boolean; set to true to not do the final processing.
+        :return: Void
+        """
+        pass
+
+    def ii_init(self, record_info_in):
+        """
+        Called when the incoming connection's record metadata is available or has changed, and
+        has let the Alteryx engine know what its output will look like.
+        :param record_info_in: XML representation for the incoming connection's field and sort properties.
+        :return: True
+        """
+        self.record_info_in = record_info_in
+        self.record_info_out = self.record_info_in
+        self.output_anchor.init(self.record_info_out)
+        self.record_creator = self.record_info_out.construct_record_creator()
+        self.record_copier = Sdk.RecordCopier(self.record_info_out, self.record_info_in)
+
+        # Map each column of the input to where we want in the output
+        for idx in range(len(self.record_info_in)):
+            self.record_copier.add(idx, idx)
+        self.record_copier.done_adding()
+        self.initialized = True
+        return True
+
+    def display_message(self, message_type, message_string):
```

Comment:
Should we notate in the docstring that these functions are not part of the Interface? Same for all the other non-interface functions?